src/libraries/System.Private.CoreLib/src/System/Numerics/BFloat16.cs

```diff
@@ -0,0 +1,2041 @@
+        /// <param name="charsWritten">When this method returns, the number of characters that were written in <paramref name="destination"/>.</param>
+        /// <param name="format">A span containing the characters that represent a standard or custom format string that defines the acceptable format for <paramref name="destination"/>.</param>
+        /// <param name="provider">An optional object that supplies culture-specific formatting information for <paramref name="destination"/>.</param>
+        /// <returns></returns>
+        public bool TryFormat(Span<char> destination, out int charsWritten, [StringSyntax(StringSyntaxAttribute.NumericFormat)] ReadOnlySpan<char> format = default, IFormatProvider? provider = null)
+        {
+            return Number.TryFormatFloat(this, format, NumberFormatInfo.GetInstance(provider), destination, out charsWritten);
+        }
+
+        /// <inheritdoc cref="IUtf8SpanFormattable.TryFormat" />
+        public bool TryFormat(Span<byte> utf8Destination, out int bytesWritten, [StringSyntax(StringSyntaxAttribute.NumericFormat)] ReadOnlySpan<char> format = default, IFormatProvider? provider = null)
+        {
+            return Number.TryFormatFloat(this, format, NumberFormatInfo.GetInstance(provider), utf8Destination, out bytesWritten);
+        }
+
+        //
+        // Explicit Convert To BFloat16
+        //
+
+        /// <summary>Explicitly converts a <see cref="char" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(char value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="decimal" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(decimal value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="double" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(double value)
+        {
+            // See explaination of the algorithm at Half.operator Half(float)
+
+            // Minimum exponent for rounding
+            const ulong MinExp = 0x3810_0000_0000_0000u;
+            // Exponent displacement #1
+            const ulong Exponent942 = 0x3ae0_0000_0000_0000u;
```

Comment:
/cc @MineCake147E I figured out the value of this magic number for double->BFloat16 by debugging, but can't give an expression to calculate it. Do you have any information around this?

Reply:
`Exponent45` reflects the difference of the number of fraction bits. `bfloat16` has 7 fraction bits, while `double` has 52. So, `52 - 7 = 45`.
`Exponent942` is a little bit complicated.
We want `1.0` to be converted to `1.0(bf16)`.

The line:

https://github.com/dotnet/runtime/blob/14b0d854392ceae0241a89c452508e3c68363e1a/src/libraries/System.Private.CoreLib/src/System/Numerics/BFloat16.cs#L490

transforms 
`0b0_011_1111_1111_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000` into `0b0_100_0010_1100_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_1000_0000`.
If we skip subtracting `Exponent942`, the `ulong newExponent = bitValue >> 45;` will be `0b0010_0001_0110_0000_0000`
By adding the value before shifting, we get `0b010000101100000000000000000000000000000000000010_0_001_0110_1_000_0000`
The internal representation of `1.0(bf16)` is `0b0_011_1111_1_000_0000`.
`0b010000101100000000000000000000000000000000000010_0_001_0110_1_000_0000 - 0b0_011_1111_1_000_0000 = 0b010000101100000000000000000000000000000000000_0011_1010_1110_0000_000`
And we shift this value 45 bits left, we get `0b0011_1010_1110_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000`, which is `0x3AE0_0000_0000_0000`.
In other words,
${ExponentDisplacement_1} = E_{source}(1.0) - E_{result}(1.0) + (FractionBits_{source} - FractionBits_{result}) + 1$
For `single` -> `half`, $127 - 15 + (23 - 10) + 1 = 113 + 13 = 126$.
For `double` -> `bfloat16`, $1023 - 127 + (52 - 7) + 1 = 897 + 45 = 942$.
We can test this formula by applying this to `double` -> `single`, $1023 - 127 + (52 - 23) + 1 = 897 + 29 = 926$.
```csharp
const ulong SingleBiasedExponentMask = double.BiasedExponentMask;
const ulong Exponent926 = 0x39e0_0000_0000_0000u;
const ulong Exponent29 = 0x01D0_0000_0000_0000u;
var q = BitConverter.DoubleToUInt64Bits(Math.PI + BitConverter.UInt64BitsToDouble(Exponent29 +
    (BitConverter.DoubleToUInt64Bits(Math.PI) & SingleBiasedExponentMask))) - Exponent926;
BitConverter.UInt32BitsToSingle((uint)(q + (q >> 29)))
```
And we get `3.1415927` which matches `(float)Math.PI`.

Reply:
It's really verbose and I haven't found best way to comment about it. Presumably I could simplify it into a simple formula like `IBinaryFloatParseAndFormatInfo` numbers.